pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/CompareObjectsWithEquals.xml

```diff
@@ -207,6 +236,79 @@ public class CompareWithEqualsTest {
     {
         return a == b;
     }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>#2934 False negative - class with fields</description>
+        <expected-problems>4</expected-problems>
+        <expected-linenumbers>8,9,10,11</expected-linenumbers>
+        <code><![CDATA[
+package net.sourceforge.pmd.lang.java.rule.errorprone.compareobjectswithequals;
+
+public class ClassWithFields {
+    private Object a;
+    private Object b;
+
+    boolean test1() {
+        return  a == b // violation
+             && this.a == b // violation
+             && a != this.b // violation
+             && this.a == this.b // violation
+             && a.equals(b) // ok
+             && this.a.equals(this.b); // ok
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>#2934 comparison with null is valid and with primitive literals</description>
+        <expected-problems>2</expected-problems>
+        <expected-linenumbers>18,19</expected-linenumbers>
+        <code><![CDATA[
+package net.sourceforge.pmd.lang.java.rule.errorprone.compareobjectswithequals;
+
+public class ClassWithFields {
+    private Object a;
+    private Object b;
+
+    boolean test1() {
+        return a == null // ok
+             && this.a == null; // ok
+    }
+
+    void test2(StringBuilder sb) {
+        if (sb != null) { } // ok
+    }
+
+    public void bar(Integer x, Integer y) {
+        if (x == 1) { } // ok
+        if (x == y) { } // not ok! only works -128<x<127 dependent on Integer caching
+        if (x == "String") { } // not ok
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>#2934 this and class should be ignored</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+package net.sourceforge.pmd.lang.java.rule.errorprone.compareobjectswithequals;
+
+public class ClassWithFields {
+    private Object a;
+    private Object b;
+
+    void test(Object o) {
+        if (this == o) { } // should be allowed, since this is a often used pattern in Object::equals.
```

Comment:
In #2899 I considered this legal only in equals method:

https://github.com/pmd/pmd/pull/2899/files#r532047509

Should we do that on master too?

Reply:
Hm... good question. I think, I considered "this" similar to "null", and didn't think longer about it (null kind of a keyword like this).

Maybe we should really restrict that to the equals method... Like I've written in the comment :) Yes, let's do that.

Reply:
I've fixed that with 7f299c4f6c3aac04fe0b6ae91f1d8afef40cc536 on master.